File: NeetCode/Binary Search/ElementInRotatedArray.java
public class ElementInRotatedArray {
    public static int findElement(int[] nums, int target) {
        int left = 0;
        int right = nums.length-1;
        
        while(left <= right) {
            int middle = left + (right-left)/2;
            if(nums[middle] == target) return middle;
            
            //? Left Sorted Portion
            if(nums[left] <= nums[middle]) {
                if(nums[left] <= target && target <= nums[middle] ) {
                    right = middle - 1;
                } else {
                    left = middle + 1;
                }
            //? Right Sorted Portion
            } else {
                if(nums[middle] <= target && target <= nums[right]) {
                    left = middle + 1;
                } else {
                    right = middle - 1;
                }
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        int[] nums = {4, 5, 6, 7, 0, 1, 2};
        int target = 3;

        System.out.println(findElement(nums, target));
    }
}

File: NeetCode/Binary Search/ElementInRotatedArray.cs
public class Solution
{
    public int Search(int[] nums, int target)
    {
        int left = 0;
        int right = nums.Length - 1;

        while (left < right)
        {
            int mid = left + (right - left) / 2;

            if (nums[mid] > nums[right])
            {
                left = mid + 1;
            }
            else
            {
                right = mid;
            }
        }

        int pivot = left;

        int res = BinarySearch(0, pivot - 1, nums, target);

        if (res != -1)
            return res;

        return BinarySearch(pivot, nums.Length - 1, nums, target);
    }

    private int BinarySearch(int left, int right, int[] nums, int target)
    {
        while (left <= right)
        {
            int mid = left + (right - left) / 2;

            if (nums[mid] == target)
            {
                return mid;
            }
            else if (nums[mid] > target)
            {
                right = mid - 1;
            }
            else
            {
                left = mid + 1;
            }
        }

        return -1;
    }
}
